dnscache/manager.go

```diff
@@ -0,0 +1,162 @@
+// Wrapped by IDnsCacheManager/DnsCacheManager. Implemented by DnsCacheStorage
+type IDnsCacheStorage interface {
+	FetchItem(key string) ([]string, error)
+	Get(key string) (DnsCacheItem, bool)
+	Set(key string, addrs []string)
+	Delete(key string)
+	Clear()
+}
+
+// DnsCacheManager is responsible for in-memory dns query records cache.
+// It allows to init dns caching and to hook into net/http dns resolution chain in order to cache query response ip records.
+type DnsCacheManager struct {
+	cacheStorage IDnsCacheStorage
+	strategy     config.IPsHandleStrategy
+	rand         *rand.Rand
+}
+
+// NewDnsCacheManager returns new empty/non-initialized DnsCacheManager
+func NewDnsCacheManager(multipleIPsHandleStrategy config.IPsHandleStrategy) *DnsCacheManager {
+	manager := &DnsCacheManager{nil, multipleIPsHandleStrategy, nil}
+	return manager
+}
+
+func (m *DnsCacheManager) SetCacheStorage(cache IDnsCacheStorage) {
+	m.cacheStorage = cache
+}
+
+func (m *DnsCacheManager) CacheStorage() IDnsCacheStorage {
+	return m.cacheStorage
+}
+
+func (m *DnsCacheManager) IsCacheEnabled() bool {
+	return m.cacheStorage != nil
+}
+
+// WrapDialer returns wrapped version of net.Dialer#DialContext func with hooked up caching of dns queries.
+//
+// Actual dns server call occures in net.Resolver#LookupIPAddr method,
+// linked to net.Dialer instance by net.Dialer#Resolver field
+func (m *DnsCacheManager) WrapDialer(dialer *net.Dialer) DialContextFunc {
+	return func(ctx context.Context, network, address string) (net.Conn, error) {
+		return m.doCachedDial(dialer, ctx, network, address)
+	}
+}
+
+func (m *DnsCacheManager) doCachedDial(d *net.Dialer, ctx context.Context, network, address string) (net.Conn, error) {
+	safeDial := func(addr string, itemKey string) (net.Conn, error) {
+		conn, err := d.DialContext(ctx, network, addr)
+		if err != nil && itemKey != "" {
+			m.cacheStorage.Delete(itemKey)
+		}
+		return conn, err
+	}
+
+	if !m.IsCacheEnabled() {
+		return safeDial(address, "")
+	}
+
+	host, port, err := net.SplitHostPort(address)
+	if err != nil {
+		return nil, err
+	}
+
+	if ip := net.ParseIP(host); ip != nil {
+		return safeDial(address, "")
+	}
+
+	ips, err := m.cacheStorage.FetchItem(host)
+	if err != nil {
+		logger.WithError(err).WithFields(logrus.Fields{
+			"network": network,
+			"address": address,
+		}).Errorf("doCachedDial cachedStorage.FetchItem error. ips=%v", ips)
+
+		return safeDial(address, "")
+	}
+
+	if m.strategy == config.NoCacheStrategy {
+		if len(ips) > 1 {
+			m.cacheStorage.Delete(host)
+			return safeDial(ips[0]+":"+port, "")
+		}
+	}
+
+	if m.strategy == config.RandomStrategy {
+		if len(ips) > 1 {
+			ip, _ := m.getRandomIp(ips)
+			return safeDial(ip+":"+port, host)
+		}
+		return safeDial(ips[0]+":"+port, host)
+	}
+
+	return safeDial(ips[0]+":"+port, host)
+}
+
+func (m *DnsCacheManager) getRandomIp(ips []string) (string, error) {
+	if m.strategy == config.RandomStrategy {
```

Comment:
sorry for being picky, somewhat about code styling - I am a big fan of early returns like:
```
func (m *DnsCacheManager) getRandomIp(ips []string) (string, error) {
	if m.strategy != config.RandomStrategy {
            return ....
         }

        // other logic
        .....
}
```
also testing `m.strategy` for  in func `getRandomIp` and calling this func only when `m.strategy == config.RandomStrategy` looks la little awkward to me, but not big deal.

Reply:
>I am a big fan of early returns like

+1, :) 

Left a check in order to prevent memory allocation for cases, when other strategies will be involved. 
With requirements change it can removed though. 
Commit - b1aae61dc18bc7ec25ad0946c5adc341811ffb9d